src/Coordination/Changelog.cpp

```diff
@@ -68,81 +78,295 @@ Checksum computeRecordChecksum(const ChangelogRecord & record)
+            changelogs_dir,
+            new_description->prefix,
+            new_start_log_index,
+            new_start_log_index + log_file_settings.rotate_interval - 1,
+            new_description->extension);
+
+        LOG_TRACE(log, "Starting new changelog {}", new_description->path);
+        auto [it, inserted] = existing_changelogs.insert(std::make_pair(new_start_log_index, std::move(new_description)));
+
+        setFile(it->second, WriteMode::Rewrite);
+    }
+
+    void finalize()
+    {
+        if (isFileSet() && prealloc_done)
+            finalizeCurrentFile();
     }
 
 private:
+
+    void finalizeCurrentFile()
+    {
+        const auto * file_buffer = tryGetFileBuffer();
+        assert(file_buffer && prealloc_done);
+
+        assert(current_file_description);
+        // compact can delete the file and we don't need to do anything
+        if (current_file_description->deleted)
+        {
+            LOG_WARNING(log, "Log {} is already deleted", file_buffer->getFileName());
+            return;
+        }
+
+        if (log_file_settings.compress_logs)
+            compressed_buffer->finalize();
+
+        flush();
+
+        if (log_file_settings.max_size != 0)
+            ftruncate(file_buffer->getFD(), initial_file_size + file_buffer->count());
+
+        if (log_file_settings.compress_logs)
+            compressed_buffer.reset();
+        else
+            file_buf.reset();
+    }
+
     WriteBuffer & getBuffer()
     {
         if (compressed_buffer)
             return *compressed_buffer;
-        return *file_buf;
+
+        if (file_buf)
+            return *file_buf;
+
+        throw Exception(ErrorCodes::LOGICAL_ERROR, "Log writer wasn't initialized for any file");
     }
 
-    std::string filepath;
+    WriteBufferFromFile & getFileBuffer()
+    {
+        auto * file_buffer = tryGetFileBuffer();
+
+        if (!file_buffer)
+            throw Exception(ErrorCodes::LOGICAL_ERROR, "Log writer wasn't initialized for any file");
+
+        return *file_buffer;
+    }
+
+    const WriteBufferFromFile * tryGetFileBuffer() const
+    {
+        return const_cast<ChangelogWriter *>(this)->tryGetFileBuffer();
+    }
+
+    WriteBufferFromFile * tryGetFileBuffer()
+    {
+        if (compressed_buffer)
+            return dynamic_cast<WriteBufferFromFile *>(compressed_buffer->getNestedBuffer());
+
+        if (file_buf)
+            return file_buf.get();
+
+        return nullptr;
+    }
+
+    void tryPreallocateForFile()
+    {
+        if (log_file_settings.max_size == 0)
+        {
+            initial_file_size = 0;
+            prealloc_done = true;
+            return;
+        }
+
+        const auto & file_buffer = getFileBuffer();
+#ifdef OS_LINUX
+        {
+            int res = 0;
+            do
+            {
+                res = fallocate(file_buffer.getFD(), FALLOC_FL_KEEP_SIZE, 0, log_file_settings.max_size + log_file_settings.overallocate_size);
+            } while (errno == EINTR);
```

Comment:
```suggestion
            int res = -1;
            do
            {
                res = fallocate(file_buffer.getFD(), FALLOC_FL_KEEP_SIZE, 0, log_file_settings.max_size + log_file_settings.overallocate_size);
            } while (res < 0 && errno == EINTR);
```

I'm not sure if `errno` is reset on success, probably it's not